File: editors/recite-neovim/lua/recite/finite_protocol.lua
-- Finite command envelope and command-specific result shapes.  Keeping this
-- separate from the streaming watch validator makes terminal semantics easy
-- to audit and keeps the lossless lexer reusable by both protocols.
local protocol = require("recite.command_protocol")
local trace = require("recite.trace_protocol")
local diagnostic = require("recite.diagnostic_protocol")
local M = {}

local function valid_artifact(value)
  return protocol.exact(value, { "path", "size_bytes" }) and protocol.machine_path(value.path)
    and protocol.integer_in_range(value.size_bytes, "0", "18446744073709551615")
end

local function valid_finite(command, terminal, exit_code)
  if not protocol.keys(terminal, { "version", "sequence", "event", "command", "status", "exit_code", "data" }, { "invocation_id" })
    or terminal.event ~= "command.result" or (terminal.status ~= "success" and terminal.status ~= "content_diagnostics")
    or protocol.compare_integer(terminal.exit_code, exit_code) ~= 0
    or terminal.status == "success" and protocol.compare_integer(terminal.exit_code, 0) ~= 0
    or terminal.status == "content_diagnostics" and protocol.compare_integer(terminal.exit_code, 1) ~= 0 then return false end
  if not protocol.object(terminal.data) then return false end
  if command == "run" or command == "trace" then
    return terminal.status == "success" and protocol.exact(terminal.data, { "trace" }) and trace.valid(terminal.data.trace)
  end
  local valid_shape = protocol.exact(terminal.data, { "diagnostics" })
  if terminal.status == "success" and command == "compile" then valid_shape = protocol.exact(terminal.data, { "diagnostics", "artifact" }) end
  if terminal.status == "success" and command == "extract" then
    valid_shape = protocol.exact(terminal.data, { "diagnostics", "artifact" }) or protocol.exact(terminal.data, { "diagnostics", "entries" })
  end
  if not valid_shape or not protocol.array(terminal.data.diagnostics) then return false end
  for _, diagnostic_record in ipairs(terminal.data.diagnostics) do if not diagnostic.valid(diagnostic_record) then return false end end
  if terminal.status == "content_diagnostics" then return protocol.exact(terminal.data, { "diagnostics" }) end
  if command == "compile" then return valid_artifact(terminal.data.artifact) end
  if command == "extract" and terminal.data.artifact ~= nil then return valid_artifact(terminal.data.artifact) end
  if command == "extract" then
    return protocol.array(terminal.data.entries) and protocol.every(terminal.data.entries, function(entry)
      return protocol.exact(entry, { "context", "source_text", "plural_source_text", "comments", "reference" })
        and type(entry.context) == "string" and type(entry.source_text) == "string"
        and (entry.plural_source_text == vim.NIL or entry.plural_source_text == nil or type(entry.plural_source_text) == "string")
        and protocol.array(entry.comments) and protocol.every(entry.comments, function(comment) return type(comment) == "string" end)
        and (entry.reference == vim.NIL or entry.reference == nil or protocol.exact(entry.reference, { "file", "line", "column" })
          and type(entry.reference.file) == "string"
          and protocol.integer_in_range(entry.reference.line, "0", "4294967295")
          and protocol.integer_in_range(entry.reference.column, "0", "4294967295"))
    end)
  end
  return true
end

function M.parse(records, command, invocation_id, exit_code)
  if not vim.tbl_contains({ "validate", "compile", "extract", "run", "trace" }, command) then error(protocol.error("unsupported_command")) end
  if #records ~= 2 then error(protocol.error("finite_record_count")) end
  local started, terminal = records[1], records[2]
  if not protocol.keys(started, { "version", "sequence", "event", "command" }, { "invocation_id" }) or started.event ~= "command.started" then error(protocol.error("missing_started")) end
  protocol.validate_envelope(started, command, invocation_id, 0)
  if terminal.event == "command.result" then
    if not valid_finite(command, terminal, exit_code) then error(protocol.error("invalid_result")) end
  elseif terminal.event == "command.error" then
    if not protocol.keys(terminal, { "version", "sequence", "event", "command", "status", "exit_code", "error" }, { "invocation_id" })
      or terminal.status ~= "failure" or protocol.compare_integer(terminal.exit_code, 1) ~= 0 or exit_code ~= 1 or not protocol.valid_error(terminal.error) then error(protocol.error("invalid_error")) end
  else error(protocol.error("missing_terminal")) end
  protocol.validate_envelope(terminal, command, invocation_id, 1)
  return { records = records, terminal = terminal }
end

return M

File: editors/vscode/src/finite-protocol.js
import { NdjsonRecordParser, integerInRange, protocol, validateEnvelope } from "./command-protocol.js";
import { validDiagnosticRecord } from "./command-diagnostics.js";
import { validStructuredError } from "./watch-record-validation.js";
import { validTrace } from "./trace-record-validation.js";

const MAX_FINITE_BYTES = 32 * 1024 * 1024;

export function parseFiniteRecords(stdout, command, invocationId, exitCode) {
  if (!["validate", "compile", "extract", "run", "trace"].includes(command)) {
    throw protocol("unsupported_command");
  }
  const parser = new NdjsonRecordParser({ maxBytes: MAX_FINITE_BYTES });
  const records = parser.push(stdout);
  parser.finish();
  if (records.length !== 2) throw protocol("finite_record_count");
  validateEnvelope(records[0], command, invocationId, 0);
  validateEnvelope(records[1], command, invocationId, 1);
  if (!exactKeys(records[0], ["version", "sequence", "event", "command", "invocation_id"]) ||
      records[0].event !== "command.started") throw protocol("missing_started");
  const terminal = records[1];
  if (terminal.event === "command.result") {
    if (!exactKeys(terminal, ["version", "sequence", "event", "command", "invocation_id", "status", "exit_code", "data"])) {
      throw protocol("invalid_result");
    }
    validateResult(command, terminal, exitCode);
  } else if (terminal.event === "command.error") {
    if (!exactKeys(terminal, ["version", "sequence", "event", "command", "invocation_id", "status", "exit_code", "error"])) {
      throw protocol("invalid_error");
    }
    validateError(terminal, exitCode);
  }
  else throw protocol("missing_terminal");
  return Object.freeze({ records, terminal });
}

function validateResult(command, record, exitCode) {
  if (![ ["success", 0], ["content_diagnostics", 1] ].some(([status, code]) =>
    record.status === status && record.exit_code === code) ||
      record.exit_code !== exitCode || !record.data || typeof record.data !== "object" ||
      Array.isArray(record.data)) throw protocol("invalid_result");
  const diagnostics = record.data.diagnostics;
  if (command === "run" || command === "trace") {
    if (record.status !== "success" || !exactKeys(record.data, ["trace"]) || !validTrace(record.data.trace)) {
      throw protocol("invalid_runtime_result");
    }
    return;
  }
  if (!Array.isArray(diagnostics) || !diagnostics.every(validDiagnosticRecord)) throw protocol("invalid_diagnostics");
  if (record.status === "content_diagnostics") {
    if (!exactKeys(record.data, ["diagnostics"])) {
      throw protocol("invalid_content_result");
    }
    return;
  }
  if (command === "validate" && !exactKeys(record.data, ["diagnostics"]) ||
      command === "compile" && !exactKeys(record.data, ["diagnostics", "artifact"]) ||
      command === "extract" && !exactKeys(record.data, ["diagnostics", "artifact"]) &&
        !exactKeys(record.data, ["diagnostics", "entries"])) {
    throw protocol("invalid_result_shape");
  }
  if (command === "compile" && !validArtifact(record.data.artifact)) throw protocol("missing_artifact");
  if (command === "extract" && !validExtractData(record.data)) throw protocol("invalid_extract_result");
}

function validateError(record, exitCode) {
  if (record.status !== "failure" || record.exit_code !== 1 || exitCode !== 1 ||
      !validStructuredError(record.error)) throw protocol("invalid_error");
}

function validExtractData(data) {
  const artifact = data.artifact !== undefined;
  const entries = data.entries !== undefined;
  return artifact !== entries && (!artifact || validArtifact(data.artifact)) &&
    (!entries || Array.isArray(data.entries) && data.entries.every(validCatalogEntry));
}

function validArtifact(value) {
  return Boolean(value && typeof value === "object" && exactKeys(value, ["path", "size_bytes"]) && machinePath(value.path) &&
    integerInRange(value.size_bytes, "0", "18446744073709551615"));
}

function validCatalogEntry(value) {
  return Boolean(value && typeof value === "object" && exactKeys(value, ["context", "source_text", "plural_source_text", "comments", "reference"]) && string(value.context) &&
    string(value.source_text) && (value.plural_source_text === null ||
      typeof value.plural_source_text === "string") &&
    Array.isArray(value.comments) && value.comments.every((comment) => typeof comment === "string") &&
    (value.reference === null || validCatalogReference(value.reference)));
}

function validCatalogReference(value) {
  return Boolean(value && typeof value === "object" && exactKeys(value, ["file", "line", "column"]) && string(value.file) &&
    Number.isSafeInteger(value.line) && value.line > 0 &&
    Number.isSafeInteger(value.column) && value.column > 0);
}

function exactKeys(value, expected) {
  const keys = Object.keys(value).sort();
  return JSON.stringify(keys) === JSON.stringify(expected.slice().sort());
}

function machinePath(value) {
  return Boolean(value && typeof value === "object" && exactKeys(value, ["encoding", "value"]) &&
    ((value.encoding === "utf8" && typeof value.value === "string") ||
      (value.encoding === "unix_bytes" && typeof value.value === "string" &&
        /^[0-9a-f]*$/u.test(value.value) && value.value.length % 2 === 0) ||
      (value.encoding === "windows_wtf16" && Array.isArray(value.value) &&
        value.value.every((unit) => Number.isSafeInteger(unit) && unit >= 0 && unit <= 0xffff))));
}

function string(value) {
  return typeof value === "string" && value.length > 0;
}
